[decision] Buffer structs (#4683)
ElektraInitiative/libelektra: doc/decisions/0_drafts/buffer_struct.md

# Buffer Struct

> ## Problem

Author: > **Note**: The problem section is very sparse, because there isn't really a huge problem here. This is more of a proposal than decision for a solution to a problem.

Reviewer: Especially abstractions should have a use case and a clear problem. E.g. avoiding code duplication is a common use case for abstraction. If it doesn't solve a problem we definitely shouldn't do it.

Author: Can you please just read my proposal and give an opinion instead of wasting both our time with criticising the process.... Why do you think I added this note when I created the PR? I know this doesn't fit the decision process. But if I had created a proposal issue, you'd have complained as well...

You can always invent some problem or come up with some abstract or even hypothetical use case. I even tried to come up with something instead of just leaving the section empty. You still found a reason to ignore the proposal.

Reviewer: I already read the problem several times: it is simply not convincing that anything should be done. I would concentrate our efforts on something that brings benefit for Elektra's user and not do internal restructuring just to have a more uniform API, i.e. less internal functions (which is currently the only reason given). Having a more generic internal API/less functions by itself **isn't better**, hence no problem and hence nothing to do. It would be very different if you want to improve the public API...

I already gave you several hints which problems might be worth solving with buffers.  If you cannot add any problem in the problem section, rejecting it as the obvious choice. More comments or notes will not help as this does not lead anywhere. Please don't reply unless you have **questions** how to proceed.

Please let us focus on problems we all agree exist, there are more than enough of those. We should be happy if something works well and no extra work is needed, shouldn't we?

Btw. in any case I appreciate that you wrote this, even if we reject it as there is **currently** no immanent benefit. Then we have documented how we could introduce buffers if they are needed at some time (e.g. public API is clumsy to use).

Author: > I already read the problem several times

**Did you read the rest?**

I know this is not the protocol for a decision. But I clearly stated that this is more of a proposal and doesn't fit the "we have a problem, let's decide a solution" process...

> It would be very different if you want to improve the public API...

**That is a goal here**

There are some examples further down in the file.... I just didn't want to design a whole public API based on a public buffer struct, only to find out, e.g. "no we don't want public structs".

> Then we have documented how we could introduce buffers if they are needed at some time (e.g. public API is clumsy to use).

Why do you think I wrote this? The current API is okay, but with #4715 anything that needs a keyname will become pretty annoying, if we need to pass a size everywhere.

> Please don't reply unless you have questions how to proceed.

Okay, I have a question: How can I convince you to be less stubborn and get you to just spend maybe five minutes reading the whole file?

For this decision a buffer is a fixed-length sequence of arbitrary bytes.
In other words, it is a pointer to data combined with the size of this data.

> Many parts of Elektra use such buffers, but there is currently no uniform API for all of these.

Reviewer: This by itself is not a problem. We should not provide an API for everything (e.g. strings, float serialization, etc.) but only for our use cases #4686.

But I agree that this area has problems and improvements are possible. Please specify what you aim for:

- optimizations? the size could be part of the allocated buffer, then it wouldn't need space if the buffer is NULL.
  Also the size field could grow as needed (with a bit indicating if further bytes also hold the size).
- less bugs? buffer resizing etc. is error-prone and we basically have no tests in this area

Also the urgency should be described. I assume it would be good to have for COW? @atmaxinger what are your thoughts on that?

Author: I know this is not really a problem (see https://github.com/ElektraInitiative/libelektra/pull/4683#discussion_r1023212527). I would have just made a proposal issue, but I wanted to avoid the inevitable "make a decision PR, please".

I have to ask, have you read the rest of the file, or are you waiting for a "clear problem"? If you haven't read the file, I would really encourage you to do so. There isn't really a clear problem here, but I think having the `ElektraBuffer` struct would make many things better.

I think when you read the whole document and see how much effort is put into the different options, it also becomes much more obvious that this is just a proposal.

> atmaxinger what are your thoughts on that?

He already left a review, see https://github.com/ElektraInitiative/libelektra/pull/4683#pullrequestreview-1181540342

Reviewer 2: > the size could be part of the allocated buffer

Honestly I think this adds a lot of complexity for a very minor gain. 

> buffer resizing etc. is error-prone 

I don't see how this would make it any better. It just groups the size of the buffer with the buffer itself. Maybe we could provide reusable methods for buffer resizing for this, then the logic only has to be implemented and tested once. But that would require that all usecases for buffer resizing are the same.

> I assume it would be good to have for COW

I'm not sure whether we gain much. Functionality wise, it has not impact whatsoever. The only way it would matter if theses buffer structs would also have reference counting, then we could just use those. But as @kodebach mentioned, reference counting is orthogonal to this. So I don't see a big correlation between this an COW.

Author: I tried to ignore the ideas for "what you aim for", because there is no goal, but some further comments:

1. Size as part of buffer: _Always_ doing this, is absolutely **NOT** a goal. It completely breaks many of the use cases I had in mind with the public struct.
    However, there are options, e.g.:
    ```c
    ElektraBuffer * elektraBufferAlloc(size_t size) {
        if (size == 0) return NULL;
        ElektraBuffer * buffer = malloc (sizeof (ElektraBuffer) + size);
        buffer->size = size;
        buffer->data = buffer + 1; // pointer arithmetic in terms of sizeof(ElektraBuffer)
        return buffer;
    }
    ```
    I _intentionally_ did not mention this in the file, because buffers created like this are a special use case of `ElektraBuffer` and providing such an API would just lead to confusion. However, for COW -- because it would all be private API -- we could do _roughly_:

    ```c
    typedef struct {
       ElektraBuffer buffer;
       uint16_t refcount;
    } RcBuffer;

    static RcBuffer * rcBufferAlloc(size_t size) {
        if (size == 0) return NULL;
        RcBuffer * rc = elektraMalloc (sizeof (RcBuffer) + size);
        rc->buffer.size = size;
        rc->buffer.data = rc + 1; // pointer arithmetic in terms of sizeof(RcBuffer)
        rc->refcount = 1;
        return rc;
    }
    
    void keySetName (Key * key, void * name, size_t size) {
        rcBufferRelease(key->name);
        key->name = allocRcBuffer(size);
        memcpy (key->name->buffer.data, name, size);
    }
    ```
2. Buffer resizing: Again not intentionally omitted. Private allocation/resizing APIs using `ElektraBuffer` are of course possible. But for a public API this could be confusing, because with the public struct this is possible:
   ```c
    char data[] = "some data";
    ElektraBuffer buffer = {.data = data, .size = sizeof(data)};
   ```
   In such a case using a resizing API `resizeBuffer(&buffer)` would obviously fail. The only way I can see having such things as public API would be with e.g.
   ```c
   typedef ElektraBuffer ElektraHeapBuffer;
   ```

3. COW: As stated, there are some ways of making the internal private APIs simpler and maybe removing some duplication. But I agree that this can definitely be done after COW is done and we shouldn't delay COW because of this decision.

Author: > This by itself is not a problem. We should not provide an API for everything

I've just read our comment again. Maybe this will help, if it does, please confirm:

```suggestion
Many parts of Elektra use such buffers, but there is currently no uniform way of using them throughout the codebase.
```

To clarify further (if I have to, I'll add something like this to the file too):
- With API I **did not** mean public API. If I did I would have written "public API".
- Yes, I do suggest possible uses in the public API and I believe using these structs in the public API would be good. But that is not what this decision is about. Stating a problem for that would have been much simpler. This decision is just about the introduction of a uniform way of handling buffers, because currently we handle buffers in whatever way the author of a particular piece of code felt like.
- This is not just about string-like buffers. It's more about the fact that C doesn't remember the size of an array. The decision could also be applied to the `Key *` array in a `KeySet`. Whether or not that would be a good idea, is a different question. But there are enough string-like buffers that the decision still makes sense without it.

## Constraints

## Assumptions

## Considered Alternatives

### Public Buffer struct

Introduce a new struct that is part of the public API and therefore declared in a public header.
The struct should be usable everywhere, so it should be declared by `libelektra-core`.

```c
typedef struct {
    void * data;
    size_t size;
} ElektraBuffer;
```

> **Note**: It is not clear yet, whether `void * data` or `const void * data` is preferrable.

This struct can be embedded into other structs, e.g.:

```c
struct _Key
{
	ElektraBuffer name;
    ElektraBuffer value;

    KeySet * meta;
	keyflag_t flags;
	uint16_t refs;
};
```

The public API for accessing the data could remain unchanged:

```c
const char * keyName (ElektraKey * key) {
    return key->name.data;
}

> const char * keyNameSize (ElektraKey * key) {

Reviewer 2: ```suggestion
const size_t keyNameSize (ElektraKey * key) {
```

    return key->name.size;
}
```

It could also return the buffer in a single function.
Either by value:

```c
const ElektraBuffer keyName (ElektraKey * key) {
    return key->name;
}

void foo(ElektraKey * key) {
    const ElektraBuffer name = keyName (key);
    printf("size %zd, first part: %s", name.size, (const char*)name.data);
}
```

or as a pointer:

```c
const ElektraBuffer * keyName (ElektraKey * key) {
    return &key->name;
}

void foo(ElektraKey * key) {
    const ElektraBuffer * name = keyName (key);
    printf("size %zd, first part: %s", name->size, (const char*)name->data);
}
```

A [short investigation](https://godbolt.org/z/r8q7YfqaW) into the generated assembly code indicates that the pointer version is likely superior.

The buffer struct could also be used as an argument in the public API.

> For simple functions that would only take one such `ElektraBuffer` argument this there is little benefit.

Reviewer 2: ```suggestion
For simple functions that would only take one such `ElektraBuffer` argument there is little benefit.
```

Especially on the caller side, the code becomes less readable:

```c
void fooNoBuffer(void * data, size_t size);
void fooWithBuffer(ElektraBuffer buffer);
void fooWithBufferPtr(ElektraBuffer * buffer);

void caller(void) {
    size_t size1 = 42;
    void * data1 = malloc(size1);

    fooNoBuffer(data1, size1);
    fooWithBuffer((ElektraBuffer){.data = data1, .size = size1});
    fooWithBufferPtr(&(ElektraBuffer){.data = data1, .size = size1});

    ElektraBuffer buffer; // buffer from somewhere else
    fooNoBuffer(buffer.data, buffer.size);
    fooWithBuffer(buffer);

>     fooWithBufferPtr(&buffer);

Reviewer 2: Not sure about this. We have to copy the buffer internally or else the data is corrupt when `buffer` goes out of scope.

Author: Totally depends on what `fooWithBufferPtr` does. Take this for example:

```c
void fooWithBufferPtr(ElektraBuffer * buffer) {
    buffer->size = 0;
}
```

Not sure when this could be used, but it is totally possible to have a "correct" `fooWithBufferPtr`.

Of course a `keySetName` with a buffer would need to make a copy. But `keySetName` already makes a copy, so that's nothing new.

The example is intentionally vague, because the main point was showing that `fooNoBuffer` is clearer.
